CalibTracker/SiStripESProducers/plugins/real/SiStripQualityESProducer.cc

```diff
@@ -16,92 +16,153 @@
 //
 //
 
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "FWCore/Framework/interface/ESProducer.h"
+#include "FWCore/Framework/interface/ModuleFactory.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DataFormats/SiStripCommon/interface/SiStripConstants.h"
+#include "CalibFormats/SiStripObjects/interface/SiStripQuality.h"
+#include "CalibTracker/Records/interface/SiStripDependentRecords.h"
+
+namespace {
+  template <typename Product, typename InputRecord>
+  class ProductGetter {
+  public:
+    virtual ~ProductGetter() = default;
+    virtual const Product& get(const InputRecord& iRecord) const = 0;
+  };
+
+  template <typename Product, typename InputRecord, typename RealRecord>
+  class ProductGetterT: public ProductGetter<Product, InputRecord> {
+  public:
+    ProductGetterT(edm::ESConsumesCollector& cc, const std::string& label):
+      token_{cc.consumesFrom<Product, RealRecord>(edm::ESInputTag{"", label})}
+    {}
+    const Product& get(const InputRecord& iRecord) const override {
+      return iRecord.get(token_);
+    }
 
-
-#include "CalibTracker/SiStripESProducers/plugins/real/SiStripQualityESProducer.h"
-
+  private:
+    edm::ESGetToken<Product, RealRecord> token_;
+  };
 
 
-SiStripQualityESProducer::SiStripQualityESProducer(const edm::ParameterSet& iConfig):
-  pset_(iConfig),
-  toGet(iConfig.getParameter<Parameters>("ListOfRecordToMerge"))
-{
-  
-  setWhatProduced(this);
-  
-  edm::LogInfo("SiStripQualityESProducer") << "ctor" << std::endl;
-
+  template <typename RealRecord>
+  auto make_badStripGetter(edm::ESConsumesCollector& cc, const std::string& label) {
+    return std::make_unique<ProductGetterT<SiStripBadStrip, SiStripQualityRcd, RealRecord> >(cc, label);
+  }
 }
 
 
-std::unique_ptr<SiStripQuality> SiStripQualityESProducer::produce(const SiStripQualityRcd& iRecord)
+class SiStripQualityESProducer : public edm::ESProducer {
+ public:
+  SiStripQualityESProducer(const edm::ParameterSet&);
+  ~SiStripQualityESProducer() override{};
+  
+  std::unique_ptr<SiStripQuality> produce(const SiStripQualityRcd&);
+   
+ private:
+  std::vector<std::unique_ptr<ProductGetter<SiStripBadStrip, SiStripQualityRcd>>> badStripGetters_;
+  // Can there be more than one entry for cabling? If not, this could be changed to a single token
+  std::vector<ProductGetterT<SiStripDetCabling, SiStripQualityRcd, SiStripDetCablingRcd>> cablingGetters_;
+  // Can there be more than one entry for VOff? If not, this could be changed to a single token
+  std::vector<ProductGetterT<SiStripDetVOff, SiStripQualityRcd, SiStripDetVOffRcd>> voffGetters_;
```

Comment:
Can there be more than one entry for cabling or VOff? In principle the previous code allowed that, so I allowed it here as well. If not, these vectors could be replaced with single `edm::ESGetToken`s.

Reply:
We don't support multiple cabling maps or VOff records at the moment, though I must admit I am not sure about the original intentions of the author about the potentiality of the code. I will have to check to give you an answer.

Reply:
> Can there be more than one entry for cabling or VOff? In principle the previous code allowed that, so I allowed it here as well. If not, these vectors could be replaced with single `edm::ESGetToken`s.

With the last commit this question is no longer relevant.